### PR TITLE
remove eachbatch and rebase DataLoader on eachobs

### PR DESCRIPTION
**Pull request by CarloLucibello**
With this, we still have a Flux compliant `DataLoader` but now the implementation is based on `eachobs`.

Also removing `eachbatch` in favor of   `eachobs(data; batchsize)`

Related to #33 #30

### CODE REVIEW

**Comment by CarloLucibello** [comment]
@lorenzoh is this implementation of the DataLoader going in the right direction?

**Comment by lorenzoh** [comment]
I think `DataLoader` doesn't need to be its own `struct`, but rather should be a wrapper that composes a data _container_ `data` with a _batch view_ `batchview([collated])` and a data _iterator_ `eachobs([parallel, buffered])`.

These 3 things should be decoupled, so the data _iterator_ `eachobs` should not handle anything related to batches.

As a sketch of `DataLoader`:

```julia
function DataLoader(data, batchsize = 1; <batchkwargs...>, <iterkwargs...>)
    return eachobs(batchview(data, batchsize; batchkwargs...), iterkwargs...)
end
```

With
- `batchkwargs`:
  - `partial = true`
  - `collate = false`
- `iterkwargs`:
  - `parallel = true`
  - `buffer = false`

I'd prefer this design as it is simpler and decouples containers, iterators and batch views.

**Comment by CarloLucibello** [comment]
Shuffling would be problematic though, as the observations are reshuffled at the first iteration each time

**Comment by lorenzoh** [comment]
You mean they should be reshuffled at every iteration or they should not?

**Comment by CarloLucibello** [comment]
at every *first* iteration
```julia
julia> d = DataLoader([1:10;], shuffle=true)
DataLoader{Vector{Int64}, Random._GLOBAL_RNG}([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 1, 10, true, true, Random._GLOBAL_RNG())

julia> collect(d)
10-element Vector{Vector{Int64}}:
 [9]
 [1]
 [2]
 [7]
 [3]
 [5]
 [8]
 [6]
 [4]
 [10]

julia> collect(d)
10-element Vector{Vector{Int64}}:
 [8]
 [9]
 [1]
 [6]
 [10]
 [4]
 [3]
 [5]
 [7]
 [2]

```

**Comment by lorenzoh** [comment]
Makes sense. In that case, let's add a `shuffle` argument to `eachobs`, what do you think?

**Comment by CarloLucibello** [comment]
pushing it down to `eachobs` means I would have to create a corresponding `EachObs` type. The problem is the same: I don't know how to achieve that shuffle behavior by just composing generators (and maybe it cannot be done). 
Anyways, defining a very simple DataLoader type as in this PR doesn't seem problematic, right?

**Comment by CarloLucibello** [comment]
merging this, we can keep iterating the implementation in other PRs
